File: AwsCryptographicMaterialProviders/runtimes/java/src/test/LocalCMCTests.java
import Random_Compile.ExternRandom;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.time.Instant;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Random;
import org.testng.annotations.Test;
import software.amazon.cryptography.keystore.model.BeaconKeyMaterials;
import software.amazon.cryptography.materialproviders.ICryptographicMaterialsCache;
import software.amazon.cryptography.materialproviders.MaterialProviders;
import software.amazon.cryptography.materialproviders.model.CacheType;
import software.amazon.cryptography.materialproviders.model.CreateCryptographicMaterialsCacheInput;
import software.amazon.cryptography.materialproviders.model.DefaultCache;
import software.amazon.cryptography.materialproviders.model.EntryDoesNotExist;
import software.amazon.cryptography.materialproviders.model.GetCacheEntryInput;
import software.amazon.cryptography.materialproviders.model.GetCacheEntryOutput;
import software.amazon.cryptography.materialproviders.model.MaterialProvidersConfig;
import software.amazon.cryptography.materialproviders.model.Materials;
import software.amazon.cryptography.materialproviders.model.PutCacheEntryInput;

public class LocalCMCTests {

  private static final ICryptographicMaterialsCache test = MaterialProviders
    .builder()
    .MaterialProvidersConfig(MaterialProvidersConfig.builder().build())
    .build()
    .CreateCryptographicMaterialsCache(
      CreateCryptographicMaterialsCacheInput
        .builder()
        .cache(
          CacheType
            .builder()
            .Default(DefaultCache.builder().entryCapacity(10).build())
            .build()
        )
        .build()
    );
  private static final List<String> identifies = Collections.unmodifiableList(
    Arrays.asList(
      "one",
      "two",
      "three",
      "four",
      "five",
      "six",
      "seven",
      "eight",
      "nine",
      "ten",
      "eleven",
      "twelve",
      "thirteen",
      "fourteen",
      "fifteen",
      "sixteen",
      "seventeen",
      "eighteen",
      "nineteen",
      "twenty",
      "twenty one"
    )
  );
  private static final int IDS_SIZE = identifies.size();

  @Test(threadPoolSize = 10, invocationCount = 300000, timeOut = 10000)
  public void TestALotOfAdding() {
    Random rand = ExternRandom.getSecureRandom();
    String beaconKeyIdentifier = identifies.get(rand.nextInt(IDS_SIZE));

    ByteBuffer cacheIdentifier = ByteBuffer.wrap(
      beaconKeyIdentifier.getBytes(StandardCharsets.UTF_8)
    );

    GetCacheEntryInput getCacheEntryInput = GetCacheEntryInput
      .builder()
      .identifier(cacheIdentifier)
      .build();

    try {
      GetCacheEntryOutput getCacheEntryOutput = test.GetCacheEntry(
        getCacheEntryInput
      );
      //      assertEquals(getCacheEntryOutput.materials().BeaconKey().beaconKey(), binaryData);
      //      assertEquals(getCacheEntryOutput.materials().BeaconKey().beaconKeyIdentifier(),
      // stringData);
      //      System.out.println("are equal");
    } catch (EntryDoesNotExist ex) {
      Materials materials = Materials
        .builder()
        .BeaconKey(
          BeaconKeyMaterials
            .builder()
            .beaconKeyIdentifier(beaconKeyIdentifier)
            // The cacheIdentifier is used as the material
            // because we are not testing the cryptography here.
            .beaconKey(cacheIdentifier)
            .encryptionContext(new HashMap<String, String>())
            .build()
        )
        .build();

      PutCacheEntryInput putCacheEntryInput = PutCacheEntryInput
        .builder()
        .identifier(cacheIdentifier)
        .creationTime(Instant.now().getEpochSecond())
        .expiryTime(Instant.now().getEpochSecond() + 1)
        .materials(materials)
        .build();
      test.PutCacheEntry(putCacheEntryInput);
    }
  }
}

File: AwsCryptographicMaterialProviders/runtimes/net/tests/LocalCMC.Tests/LocalCMCTests.cs
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Text;
using AWS.Cryptography.KeyStore;
using AWS.Cryptography.MaterialProviders;
using Org.BouncyCastle.Security;
using Random_Compile;
using Xunit;


public class LocalCMCTests
{
  private static MaterialProviders mpl = new MaterialProviders(new MaterialProvidersConfig());
  private static ICryptographicMaterialsCache test = mpl.CreateCryptographicMaterialsCache(
      new CreateCryptographicMaterialsCacheInput { Cache = new CacheType{Default = new DefaultCache{EntryCapacity = 10}}}
    );
  
  private static ReadOnlyCollection<string> identifiers = 
    new ReadOnlyCollection<string>(new List<string>{
      "one",
      "two",
      "three",
      "four",
      "five",
      "six",
      "seven",
      "eight",
      "nine",
      "ten",
      "eleven",
      "twelve",
      "thirteen",
      "fourteen",
      "fifteen",
      "sixteen",
      "seventeen",
      "eighteen",
      "nineteen",
      "twenty",
      "twenty one"}
    );

  private static int ID_SIZE = identifiers.Count;

  [Fact]
  public void TestLotsOfAdding()
  {
    int threadCount = 10;
    int numTasks = 300000;
    
    ThreadPool.SetMaxThreads(threadCount, threadCount);
    for (int i = 0; i < numTasks; i++)
    {
      ThreadPool.QueueUserWorkItem(WorkForThread);
    }
  }

  private static void WorkForThread(object? state)
  {
    Console.WriteLine("TestLotsOfAdd");
    Random random = ThreadSafeRandom.getSecureRandom();
    var beaconKeyIdentifier = identifiers[random.Next(ID_SIZE)];

    MemoryStream cacheIdentifier = new MemoryStream(Encoding.UTF8.GetBytes(beaconKeyIdentifier));
    GetCacheEntryInput cacheEntryInput = new GetCacheEntryInput{Identifier = cacheIdentifier};

    try
    {
      GetCacheEntryOutput cacheEntryOutput = test.GetCacheEntry(cacheEntryInput);
    }
    catch (EntryDoesNotExist e)
    {
      Materials materials = new Materials
      {
        BeaconKey = new BeaconKeyMaterials
        {
          BeaconKeyIdentifier = beaconKeyIdentifier,
          BeaconKey = cacheIdentifier,
          EncryptionContext = new Dictionary<string, string>()
        }
      };

      PutCacheEntryInput putCacheEntryInput = new PutCacheEntryInput
      {
        Identifier = cacheIdentifier,
        CreationTime = DateTimeOffset.UnixEpoch.Second,
        ExpiryTime = DateTimeOffset.UnixEpoch.Second + 1,
        Materials = materials
      };

      test.PutCacheEntry(putCacheEntryInput);
    }
  }
  
  // ONLY USED FOR TESTING
  public partial class ThreadSafeRandom
  {
      [ThreadStatic] private static SecureRandom? _local;

      private static SecureRandom Instance
      {
          get
          {
              if (_local is null)
              {
                  _local = new SecureRandom();
              }
              return _local;
          }
      }

      public static SecureRandom getSecureRandom()
      {
          return Instance;
      }
  }

}
